LabManagement/PowerShell/BulkOperations/Az.LabServices.BulkOperations.psm1

```diff
@@ -1438,6 +1438,178 @@ function Remove-AzLabUsersBulk {
     }
 }
 
+# This function is used to send lab invitations *after* a lab has been created\published - for example, several days\weeks later when student enrollment is finalized.
+function Send-AzLabsInvitationBulk {
+    param(
+        [parameter(Mandatory = $true, HelpMessage = "Array containing one line for each lab", ValueFromPipeline = $true)]
+        [ValidateNotNullOrEmpty()]
+        [psobject[]]
+        $labs,
+
+        [parameter(Mandatory = $false, ValueFromPipelineByPropertyName = $true)]
+        [int]
+        $ThrottleLimit = 5
+    )
+
+    begin {
+        # This patterns aggregates all the objects in the pipeline before performing an operation
+        # i.e. executes lab creation in parallel instead of sequentially
+        # This trick is to make it work both when the argument is passed as pipeline and as normal arg.
+        # I came up with this. Maybe there is a better way.
+        $aggregateLabs = @()
+    }
+    process {
+        # If passed through pipeline, $labs is a single object, otherwise it is the whole array
+        # It works because PS uses '+' to add objects or arrays to an array
+        $aggregateLabs += $labs
+
+    }
+    end {
+        $init = {            
+        }
+
+        function Send-AzLabsInvitation-Jobs {
+            [CmdletBinding()]
+            param(
+                [parameter(Mandatory = $true, ValueFromPipeline = $true)]
+                [psobject[]]
+                $ConfigObject
+            )
+
+            $block = {
+                param($path)
+
+                Set-StrictMode -Version Latest
+                $ErrorActionPreference = 'Stop'
+
+                Import-Module -Name Az.LabServices
+
+                $input.movenext() | Out-Null
+            
+                $obj = $input.current[0]
+            
+                if ($obj.Invitation -and $obj.Emails) { 
+                    Write-Host "Sending invitation emails for lab: $($obj.LabName)."
+                    $lab = Get-AzLabServicesLab -ResourceGroupName $obj.ResourceGroupName -Name $($obj.LabName)
+                    if (-not $lab -or @($lab).Count -ne 1) { Write-Error "Unable to find lab $($obj.LabName)." }
+
+                    $users = Get-AzLabServicesUser -LabName $lab.Name -ResourceGroupName $obj.ResourceGroupName
+                    if (($users | Measure-Object).Count -eq 0) { Write-Error "The lab doesn't have any users added." }
+
+                    foreach ($user in $users ) {
+                        Write-Verbose "Sending invitation email to user: $($user.Email)."
```

Comment:
In this code - it looks like we're sending invitations to ALL lab users, we are ignoring the emails that were passed in via the CSV file.  Is that intentional?  If so - maybe we can remove the requirement for emails on line 1491 since it's not needed?

Perhaps another approach for this method might be to send an invitation to any students who haven't yet registered?   So for any students where invitations were sent, they'd get another one, and for students where invitations weren't sent, they would get their first?

Reply:
Fixed - I check for Invitationstate == "NotSent"